### PR TITLE
feat(access): Improve external role retrieval

### PR DESCRIPTION
**Pull request by filipe-caetano-ovo**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### Summary
This PR improves the Access Management tab performance, more specifically fetching the dataset's roles and associated users + checking if the current user is assigned to them in order to display request/provisioned access button.

### Problem
Currently, the `getExternalRoles` query (performed when clicking on the AccessManagement tab) fetches all actors/users for all roles of a specific dataset and then compares those URNs to determine if the current user is assigned to each role. It then displays a Request/Provisioned access button according to this result. We were running into very slow loading times for roles with many users assigned to them (one specific dataset had a role assigned to it with ~8k users took >8s to load). This effectively makes the feature unusable past a certain scale as the user gives up before the role info load is complete.

### Solution
* A custom `isAssignedToMe` field and Resolver were added to the Role entity to perform this filtering in the backend;
* The `getExternalRoles` query and handler func were modified to use the above field to determine if current user has a role.

### CODE REVIEW

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataset/ActorUsersResolver.java
Comment (chriscollins3456) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataset/ActorUsersResolver.java:53]: so the way i'm reading this logic is that if you have a userFilter with a user urn in it, simply fetch the user and populate it in the list of users. Wouldn't this be incorrect, though, if the user is not in this role? with this filter the user you provide would always be in the list of every role you give.

I think we need a different approach to improving this performance here.

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataset/ActorUsersResolver.java
Comment (chriscollins3456) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/dataset/ActorUsersResolver.java:62]: similarly with this logic I believe we just populate the first 10,000 users from your DB as belonging to this role if there's no filter. I'll suggest an alternative approach we can take here in a separate comment

File: datahub-graphql-core/src/main/resources/entity.graphql
Comment (chriscollins3456) [inline datahub-graphql-core/src/main/resources/entity.graphql:1730]: what if, instead of putting a filter here, we create a new shortcut field for getting whether or not a given role is simply assigned to "me" (the querying user)

what that could look like is a new field on the `Role` type right above called something like `isAssignedToMe: Boolean!`

this field can then have its own custom resolver similar to what you've done above. something very similar (even with the same name I suggest above) is the `IsFormAssignedToMeResolver` for our documentation forms feature. [in GMSGraphQLEngine we have this resolver defined](https://github.com/acryldata/datahub-fork/blob/acryl-main/datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GmsGraphQLEngine.java#L2625) under the `FormActorAssignment` parent. We could do something similar but with the `Role` parent, and have a new resolver that can get the role and the requesting user and check to see if that specific user is in the role.

the logic would be very similar to what we do in `IsFormAssignedToMeResolver` but we could ignore the groups piece.

Therefore it would be much more performant since we could simply check the list from the `Role` object and see if the logged in user is in there and return a boolean yes/no. Then if we stop querying for the list of users from the frontend, we should be in good shape since we would then not hydrate all of these users.

File: datahub-web-react/src/graphql/dataset.graphql
Comment (chriscollins3456) [inline datahub-web-react/src/graphql/dataset.graphql:334]: then if we add the new `isAssignedToMe` field, we can update this query to not even fetch the `users` here since it looks like right now the UI is only getting `users` to check if the logged in user is in the list. So that would be a big performance boost

File: datahub-graphql-core/src/main/resources/entity.graphql
Comment (chriscollins3456) [reply datahub-graphql-core/src/main/resources/entity.graphql:1730]: or the field could be `userHasAccess` or something

**Review by chriscollins3456 — COMMENTED** [review]
thank you for taking this task on! I have a few concerns about your current approach but offer a suggestion for another way to handle this problem

there's definitely more than one solution here so let me know if you like my approach or if you are able to modify your current approach to work for the situation

File: datahub-graphql-core/src/main/resources/entity.graphql
Comment (filipe-caetano-ovo) [reply datahub-graphql-core/src/main/resources/entity.graphql:1730]: Hi Chris, thanks for the feedback. I am aware that the solution would not return the appropriate users, just wanted to get some feedback on what would be the best way to approach the problem seeing that I'm not super familiar with the code base. The `isAssignedToMe` suggestion sounds good (I was trying to achieve something similar, i.e.: not hydrating all the users of the role if not necessary) 👍 I'll try to get that implemented.

Comment (chriscollins3456) [reply datahub-graphql-core/src/main/resources/entity.graphql:1730]: okay sounds great thanks so much! yeah sorry I didn't realize it was more of a placeholder to start the conversation / get feedback. but it served its job well! let me know if you need any assistance or anything :)

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GmsGraphQLEngine.java
Comment (chriscollins3456) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GmsGraphQLEngine.java:2931]: small one: but maybe let's rename this to `configureRoleResolvers` and rename the current `configureRoleResolvers` to be `configureDataHubRoleResolvers`

**Review by chriscollins3456 — APPROVED** [review]
nice this looks great! I have one request for a rename of a function but not blocking my approval. If you could get the rename in that's great, otherwise I can always take care of that once this is merged when CI is green

**Comment by chriscollins3456** [comment]
@filipe-caetano-ovo do you think this still needs to be a draft PR or do you think it's ready to merge? the code looks good, and if things are working well on your side as well we should be good

**Comment by filipe-caetano-ovo** [comment]
@chriscollins3456 fine by me to merge, there are some [jobs failing](https://github.com/datahub-project/datahub/actions/runs/8597360234/job/23556273299?pr=10160) on the DockerHub login step, not sure why.

**Comment by chriscollins3456** [comment]
> @chriscollins3456 fine by me to merge, there are some [jobs failing](https://github.com/datahub-project/datahub/actions/runs/8597360234/job/23556273299?pr=10160) on the DockerHub login step, not sure why.

ah yeah we're having an issue with our CI flow right now and in the process of fixing that. i think i'll hold off on merging until that gets resolved! that should hopefully be today or at least beginning of this week

**Pull request merged**
